File: 2023/day07/day07.py
from collections import defaultdict


def str_to_val(s: str) -> int:
    if s == "A":
        return 13
    elif s == "K":
        return 12
    elif s == "Q":
        return 11
    elif s == "T":
        return 10
    elif s == "J":
        return 1
    else:
        return int(s)


class Wager:
    card: str
    value: int
    strength: int

    def __init__(self, card, value, strength) -> None:
        self.card = card
        self.value = value
        self.strength = strength

    def __lt__(self, other) -> bool:
        if self.strength != other.strength:
            return self.strength < other.strength
        
        for i in range(5):
            if self.card[i] != other.card[i]:
                return str_to_val(self.card[i]) < str_to_val(other.card[i])
            
        return True
    
    def __repr__(self) -> str:
        return "{" + f"card: {self.card}, value: {self.value}, strength: {self.strength}" + "}"


def main():
    with open("input.txt") as f:
        lines = f.readlines()

    wagers = []
    for line in lines:
        card, bet = line.split(" ")
        value = int(bet)

        counts = defaultdict(int)
        for c in card:
            counts[c] += 1

        jokers = counts.pop("J") if "J" in counts else 0
        if jokers == 5:
            wagers.append(Wager(card, value, 6))
            continue

        sets = defaultdict(int)
        for v in counts.values():
            sets[v] += 1

        print(sets)
        most = sorted(sets, reverse=True)[0]
        sets[most] -= 1
        sets[most + jokers] += 1

        strength = 0
        if sets[5] == 1:
            strength = 6
        elif sets[4] == 1:
            strength = 5
        elif sets[3] == 1 and sets[2] == 1:
            strength = 4
        elif sets[3] == 1:
            strength = 3
        elif sets[2] == 2:
            strength = 2
        elif sets[2] == 1:
            strength = 1

        wagers.append(Wager(card, value, strength))

    wagers.sort()

    print(wagers)

    total = 0
    for i, wager in enumerate(wagers, start = 1):
        total += i * wager.value

    print(total)


if __name__ == "__main__":
    main()

File: 2023/day07/day07.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type wager struct {
	cards    string
	value    int
	strength int
}

func byteToVal(b byte) int {
	switch b {
	case 'A':
		return 13
	case 'K':
		return 12
	case 'Q':
		return 11
	case 'J':
		return 10
	case 'T':
		return 9
	default:
		num, _ := strconv.Atoi(string(b))
		return num - 1
	}
}

func main() {
	body, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Printf("failed to read file: %v\n", err.Error())
		return
	}

	lines := strings.Split(strings.Trim(string(body), "\n"), "\n")
	wagers := make([]wager, 0)
	for _, line := range lines {
		components := strings.Split(line, " ")
		counts := map[rune]int{}

		for _, c := range components[0] {
			counts[c] += 1
		}

		sets := map[int]int{}
		for _, v := range counts {
			sets[v] += 1
		}

		strength := 0
		if sets[5] == 1 {
			strength = 6
		} else if sets[4] == 1 {
			strength = 5
		} else if sets[3] == 1 && sets[2] == 1 {
			strength = 4
		} else if sets[3] == 1 {
			strength = 3
		} else if sets[2] == 2 {
			strength = 2
		} else if sets[2] == 1 {
			strength = 1
		}

		value, _ := strconv.Atoi(components[1])
		wagers = append(wagers, wager{cards: components[0], value: value, strength: strength})
	}

	sort.SliceStable(wagers, func(i, j int) bool {
		a := wagers[i]
		b := wagers[j]

		if a.strength != b.strength {
			return a.strength < b.strength
		}

		for k := 0; k < 5; k++ {
			if a.cards[k] != b.cards[k] {
				return byteToVal(a.cards[k]) < byteToVal(b.cards[k])
			}
		}

		return true
	})

	fmt.Println(wagers)

	sum := 0
	for i, val := range wagers {
		sum += (i + 1) * val.value
	}

	fmt.Println(sum)
}
